03-path-application-development/310-chaos-engineering/readme.adoc

```diff
@@ -0,0 +1,281 @@
+
+== Prerequisites
+
+This chapter uses a cluster with 3 master nodes and 5 worker nodes as described here: link:../cluster-install#multi-master-multi-node-multi-az-gossip-based-cluster[multi-master, multi-node gossip based cluster].
+
+All configuration files for this chapter are in the `experiments` directory. Make sure you change to that directory before executing any commands from this chapter.
+
+You will need to link:http://chaostoolkit.org/reference/usage/install/[install the open source Chaos Toolkit] to run the examples in this chapter.
+
+== Introduction to Chaos Engineering
+
+Chaos Engineering is the discipline of working with the link:http://principlesofchaos.org/[uncertainty of distributed systems at scale]. Modern distributed systems naturally become more complex in order to accomodate essential system size, distribution and the desirable benefit of increased speed of system change.
+
+Given the number of variables involved, it is difficult in the extreme to formally, and before-time, prove that a system is faultless, especially when the system is constantly changing. Chaos Engineering is an answer to this challenge as it specifically provides a means of improving confidence in the system's availability under various production conditions.
+
+An empirical process, Chaos Engineering experiments exercise a distributed system to see what weaknesses can be found.
+
+=== The Chaos Engineering Process
+
+The link:http://principlesofchaos.org/["Principles of Chaos"] define the practical process that Chaos Engineering executes as:
+
+1. Start by defining ‘steady state’ as some measurable output of a system that indicates normal behavior.
+2. Hypothesize that this steady state will continue in both the control group and the experimental group.
+3. Introduce variables that reflect real world events like servers that crash, hard drives that malfunction, network connections that are severed, etc.
+4. Try to disprove the hypothesis by looking for a difference in steady state between the control group and the experimental group.
+
+In this chapter we will explore implementing this process using the free and open source link:http://chaostoolkit.org/[Chaos Toolkit].
+
+== Application Architecture Recap
+
+The sample application already created by the pre-requisites uses three services:
+
+[.thumb]
+image::services.png[]
+
+. `webapp`: Web application microservice uses `greeter` and `name` microservice to generate a greeting for a person.
+. `greeter`: A microservice returns a greeting based upon `greet` name/value keypair in the URL.
+. `name`: A microservice that returns person's name based upon `id` name/value keypair in the URL.
+
+These services are built as Docker images and deployed in Kubernetes. All services are built as Node.js applications. The source code for the services is at https://github.com/arun-gupta/container-service-discovery/tree/master/services.
```

Comment:
For simplicity and inclusion, we should provide the steps here to deploy the services in the k8s cluster.

Reply:
Would this be a link to the `local-deployment` instructions in 301? Or did you have something larger in mind?

Reply:
Its not about the deployment of the cluster, but deployment of the services mentioned in the github repo.